src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -0,0 +1,266 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ODataUtf8JsonWriter.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+#if NETCOREAPP3_1_OR_GREATER
+using System;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Globalization;
+using System.Text.Encodings.Web;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.OData.Json
+{
+    /// <summary>
+    /// Implement of <see cref="IJsonWriter"/> that is based on
+    /// <see cref="Utf8JsonWriter"/>.
+    /// </summary>
+    internal class ODataUtf8JsonWriter : IJsonWriter, IDisposable
+    {
+        private const int DefaultBufferSize = 16 * 1024;
+        private const float BufferFlushThreshold = 0.9f;
+        private readonly static byte[] Parentheses = new byte[] { (byte)'(', (byte)')' };
+
+        private readonly Stream outputStream;
+        private readonly Utf8JsonWriter writer;
+        private readonly int bufferSize;
+        private readonly bool isIeee754Compatible;
+        private readonly bool leaveStreamOpen;
+        private readonly Encoding encoding = Encoding.UTF8;
+        private bool disposed;
+
+        public ODataUtf8JsonWriter(Stream outputStream, bool isIeee754Compatible, int bufferSize = DefaultBufferSize, bool leaveStreamOpen = false)
+        {
+            this.outputStream = outputStream;
+            this.isIeee754Compatible = isIeee754Compatible;
+            this.bufferSize = bufferSize;
+            this.writer = new Utf8JsonWriter(
+                outputStream,
+                // we don't need to perform validation here since the higher-level
+                // writers already perform validation
+                new JsonWriterOptions { SkipValidation = true });
+            this.leaveStreamOpen = leaveStreamOpen;
+        }
+
+        public void Flush()
+        {
+            if (this.writer.BytesPending > 0)
+            {
+                this.writer.Flush();
+            }
+        }
+
+        public void FlushIfBufferThresholdReached()

```

Comment:
It seems that we currently write directly to the stream without a buffer. That can be achieved here by providing `0` as `bufferSize` in the constructor, but do we want to preserve that behavior as the default?

Reply:
In this case I think the buffering is the right thing to do since it's one of the arguments I had for adopting `Utf8JsonWriter`. I had also initially proposed adding buffering to the existing writer (e.g. by wrapping a `BufferedStream` like we do with the async writer) and that also seemed to slightly improve perf.

I think we should use `Utf8JsonWriter` the way it was intended to be used in order to maximize the perf benefits, otherwise there are fewer reasons to use it.

Reply:
> The main reason we want the customers to move to the new writer is for performance

I very strongly disagree with this. We should move to `Utf8JsonWriter` because it is the standard, and we should use standard implementations unless we have good reason not to. The performance benefits are an added bonus.

Either way, I'm not so attached to preserving the buffering behavior, so I guess this is fine.